php/libraries/User.class.inc

```diff
@@ -541,5 +541,24 @@ class User extends UserPermissions
         return $this->hasPermission($code)
             && $this->hasCenter($center_id);
     }
+
+    /**
+     * Determines if the user has ever logged in successfully
+     *
+     * @return bool
+     */
+    public function hasEverLoggedIn(): bool
+    {
+        $count = (\Database::singleton())->pselectOne(
```

Comment:
when do we \Database::singleton() and when do we \NDB_Factory::singleton(), factory->database()?

Reply:
You are right, I should change that